keymanapp/keyman
# change(mac): store partial path in UserDefaults

Rather than store the full path of keyboard files throughout the Keyman settings in the UserDefaults, simply store the relative path from the Keyman-Keyboards directory.

With the encapsulation of the settings/UserDefaults in KMSettingsRepository, this also allowed other references to the settings to access it directly rather than gaining access via the application delegate.

This is dependent on #12106 which changed the location of the Keyman data on disk. 

Fixes #2542

----

For testing purposes, the current values of Keyman's settings in the UserDefaults can be found by executing the following command in a terminal window:

`defaults read keyman.inputmethod.Keyman`

These settings can be written to a plist file in the current directory with

`defaults read keyman.inputmethod.Keyman > ./keyman-userdefaults.plist`

These settings can be imported back into the UserDefaults as follows:

`defaults import keyman.inputmethod.Keyman ./keyman-userdefaults.plist`

Other commands are also useful for testing:
delete all Keyman settings: `defaults delete keyman.inputmethod.Keyman`
delete a single setting: `defaults delete keyman.inputmethod.Keyman KMDataModelVersion`
set a single setting: `defaults write keyman.inputmethod.Keyman KMDataModelVersion 1`
get documentation: `'defaults help'`

# User Testing

## TEST_KEYBOARDS_MOVED_TO_LIBRARY

1. Use Keyman 17 for Mac and install a couple keyboards
1. Verify that these keyboards are located in ~/Documents/Keyman-Keyboards
1. Install and run the new version of Keyman for this PR
1. Verify that the keyboards files have all been moved to ~/Library/Application 

Support/keyman.inputmethod.Keyman/Keyman-Keyboards

**TEST_ARMENIAN_OPTION_ENABLED**: 

1. Open the Stickies app on the Mac and create a new note
1. Switch to the Armenian Mnemonic keyboard 
1. Type <kbd>e</kbd><kbd>w</kbd>
1. Confirm that this produces the output: եւ
1. Type <kbd>option-v</kbd><kbd>w</kbd> (this toggles a setting within Keyman for this specific keyboard)
1. Type <kbd>e</kbd><kbd>w</kbd>
1. Confirm that this same input as earlier now produces the output: և
1. In a terminal window type the command `defaults read keyman.inputmethod.Keyman`
1. Confirm that the Armenian keyboard has an option of `option_ligature_ew` with a value of `1`. 

**TEST_ARMENIAN_OPTION_DISABLED**: 

1. With the current keyboard set to Armenian Mnemonic, and after executing TEST_ARMENIAN_OPTION_ENABLED,  type <kbd>option-v</kbd><kbd>w</kbd> to disable the option `option_ligature_ew`
1. Type <kbd>e</kbd><kbd>w</kbd>
1. Confirm that this produces the output: եւ
1. In a terminal window type the command `defaults read keyman.inputmethod.Keyman`
1. Confirm that the Armenian keyboard has an option of `option_ligature_ew` with a value of `0`. 

**TEST_SELECTED_KEYBOARD**: 

1. Install the keyboard EuroLatin (SIL) and select it as the current keyboard
1. In a terminal window type the command `defaults read keyman.inputmethod.Keyman`
1. Confirm that the property `KMSelectedKeyboardKey` is set to the value `"/sil_euro_latin/sil_euro_latin.kmx"`
